File: old/salary.py
import xml.etree.ElementTree as ET
import pandas as pd

#metavariabler
FORMAT = "LÖNIN"
VERSION = "2.1"
LAND = "Sverige"
PROGRAMNAMN = "Lönekörning"
EXCEL_FILE = "./lönefiler/input/testfil2.xlsx"
OUTPUT_FILE = "./lönefiler/output/visa_david2.pax"
DATE = "2024-03-12"

#create header
def create_metatag():
    return b'<?xml version="1.0" encoding="utf-8"?>\n'

def create_header():
    header = ET.Element("header")
    format_element = ET.SubElement(header, "format")
    version_element = ET.SubElement(header, "version")
    land_element = ET.SubElement(header, "land")
    programnamn_element = ET.SubElement(header, "programnamn")
    
    format_element.text = FORMAT
    version_element.text = VERSION
    land_element.text = LAND
    programnamn_element.text = PROGRAMNAMN

    return header

#create the folding lonetransaktioner
def create_lon_folder():
    lonetransaktioner = ET.Element("lonetransaktioner")
    return lonetransaktioner

#create one lönetransaktion based on indata
def create_transaction(persnr, lonart, benamning, antal, apris, kommentar, datum):
    lonetrans = ET.Element("lonetrans")
    lonetrans.set('persnr', str(persnr))

    lonart_el = ET.SubElement(lonetrans, "lonart")
    lonart_el.text = lonart

    benamning_el = ET.SubElement(lonetrans, "benamning")
    #denna bör kanske vara automatisk
    benamning_el.text = benamning

    antal_el = ET.SubElement(lonetrans, "antal")
    antal_el.text = str(antal)

    apris_el = ET.SubElement(lonetrans, "apris")
    apris_el.text = str(apris)

    belopp_el = ET.SubElement(lonetrans, "belopp")
    belopp = antal*apris
    belopp_el.text = str(belopp)

    kommentar_el = ET.SubElement(lonetrans, "kommentar")
    kommentar_el.text = kommentar

    datum_el = ET.SubElement(lonetrans, "datum")
    datum_el.text = datum

    return lonetrans

#create around tag
def around_tag():
    paxml = ET.Element("paxml")
    paxml.set("xmlns:xsi", 'http://www.w3.org/2001/XMLSchema-instance')
    paxml.set("xsi:noNamespaceSchemaLocation", 'http://www.paxml.se/2.1/paxml.xsd')
    return paxml

#läs in excel
def read_file():
    dataframe1 = pd.read_excel(EXCEL_FILE)
    print(dataframe1)

    #initiera frame:
    parent = around_tag()
    #header
    header = create_header()
    parent.append(header)

    #där alla trans ligger
    wrapper = ET.Element("lonetransaktioner")
    parent.append(wrapper)

    for index, row in dataframe1.iterrows():
        #tar bort alla som inte ska få lön inrapporterat
        
        if row["Lön"] != 0 and row["Fast lön"] != "JA":
            personnummer = row["Personnummer"]
            lön_träning = row["Tr"]
            lön_match = row["ma"]
            lön_cup = row["Cu/lä"]
            print(f"{personnummer} {lön_träning} {lön_match} {lön_cup}")
            
            #
            totalen = 0
            #Träning
            if row["TrN"] > 0:
                salary = row["TrN"]*lön_träning
                totalen += salary
                #behöver eventuellt läggas till belopp
                lonetrans = create_transaction(personnummer, "112", "Ersättning träning (antal tillfällen)", row["TrN"], row["Tr"], "test", DATE)
                wrapper.append(lonetrans)

            #Match
            if row["MaN"] > 0:
                salary = row["MaN"]*lön_match
                totalen += salary
                lonetrans = create_transaction(personnummer, "113", "Ersättning match (antal tillfällen)", row["MaN"], row["ma"], "test", DATE)
                wrapper.append(lonetrans)

            #Cup/läger
            if row["CupN"] > 0 or row["LägerN"] > 0:
                salary = (row["CupN"] + row["LägerN"])*lön_cup
                totalen +=salary
                total_number = row["CupN"] + row["LägerN"]
                lonetrans = create_transaction(personnummer, "114", "Ersättning cup/läger (antal dagar)", total_number, row["Cu/lä"], "test", DATE)
                wrapper.append(lonetrans)
            
            #ta tillbaka eventuellt, fattar inte riktigt PAX
            """#semesterersättning på allt
            semesterers = round(totalen*0.12, 1)
            #här kanske det behövs en total
            lonetrans = create_transaction(personnummer, "224", "Semesterersättning i anslutning till lön", 1, semesterers, "test", DATE)
            wrapper.append(lonetrans)"""

    #print("kommer ändå hit")
    return parent



#main
def main():
    parent = around_tag()
    lonetrans = create_transaction("1234567890", "Lonart", "Benamning", 5, 10.5, "Kommentar", "2024-03-12")
    parent.append(lonetrans)
    return parent
    


#write to file
with open(OUTPUT_FILE, "wb") as f:
    metatag = create_metatag()
    f.write(metatag)    
    parent = read_file()
    b_parent = ET.tostring(parent)
    f.write(b_parent)

File: old/salary.js
import fs from 'node:fs';
import path from 'node:path';
import xlsx from 'xlsx';
import { create } from 'xmlbuilder2';

// Meta variables
const FORMAT = "LÖNIN";
const VERSION = "2.1";
const LAND = "Sverige";
const PROGRAMNAMN = "Lönekörning";
const EXCEL_FILE = "./lönefiler/input/testfil2.xlsx";
const OUTPUT_FILE = "./lönefiler/output/visa_david2.pax";
const DATE = "2024-03-12";

function createHeader(parent) {
    const header = parent.ele("header");
    header.ele("format").txt(FORMAT);
    header.ele("version").txt(VERSION);
    header.ele("land").txt(LAND);
    header.ele("programnamn").txt(PROGRAMNAMN);
    return header;
}

function createTransaction(wrapper, persnr, lonart, benamning, antal, apris, kommentar, datum) {
    const lonetrans = wrapper.ele("lonetrans").att("persnr", String(persnr));
    lonetrans.ele("lonart").txt(String(lonart));
    lonetrans.ele("benamning").txt(benamning);
    lonetrans.ele("antal").txt(String(antal));
    lonetrans.ele("apris").txt(String(apris));
    const belopp = Number(antal) * Number(apris);
    lonetrans.ele("belopp").txt(String(belopp));
    lonetrans.ele("kommentar").txt(kommentar);
    lonetrans.ele("datum").txt(datum);
    return lonetrans;
}

function createRoot() {
    const root = create().ele("paxml");
    root.att("xmlns:xsi", "http://www.w3.org/2001/XMLSchema-instance");
    root.att("xsi:noNamespaceSchemaLocation", "http://www.paxml.se/2.1/paxml.xsd");
    return root;
}

function readFileAndBuildXml() {
    const workbook = xlsx.readFile(EXCEL_FILE);
    const sheetName = workbook.SheetNames[0];
    const worksheet = workbook.Sheets[sheetName];
    const rows = xlsx.utils.sheet_to_json(worksheet, { raw: true });

    const parent = createRoot();
    createHeader(parent);

    const wrapper = parent.ele("lonetransaktioner");

    for (const row of rows) {
        const lon = row["Lön"];
        const fastLon = row["Fast lön"];

        if ((lon ?? 0) !== 0 && fastLon !== "JA") {
            const personnummer = row["Personnummer"];
            const lonTraning = row["Tr"];
            const lonMatch = row["ma"];
            const lonCup = row["Cu/lä"];
            // Log just like Python
            console.log(`${personnummer} ${lonTraning} ${lonMatch} ${lonCup}`);

            let totalen = 0;

            // Träning
            const trN = Number(row["TrN"] || 0);
            if (trN > 0) {
                const apris = Number(row["Tr"]);
                totalen += trN * apris;
                createTransaction(
                    wrapper,
                    personnummer,
                    "112",
                    "Ersättning träning (antal tillfällen)",
                    trN,
                    apris,
                    "test",
                    DATE
                );
            }

            // Match
            const maN = Number(row["MaN"] || 0);
            if (maN > 0) {
                const apris = Number(row["ma"]);
                totalen += maN * apris;
                createTransaction(
                    wrapper,
                    personnummer,
                    "113",
                    "Ersättning match (antal tillfällen)",
                    maN,
                    apris,
                    "test",
                    DATE
                );
            }

            // Cup/läger
            const cupN = Number(row["CupN"] || 0);
            const lagerN = Number(row["LägerN"] || 0);
            if (cupN > 0 || lagerN > 0) {
                const totalNumber = cupN + lagerN;
                const apris = Number(row["Cu/lä"]);
                totalen += totalNumber * apris;
                createTransaction(
                    wrapper,
                    personnummer,
                    "114",
                    "Ersättning cup/läger (antal dagar)",
                    totalNumber,
                    apris,
                    "test",
                    DATE
                );
            }

            // Optional (commented out in Python): semesterersättning
            // const semesterers = Math.round(totalen * 0.12 * 10) / 10;
            // createTransaction(wrapper, personnummer, "224", "Semesterersättning i anslutning till lön", 1, semesterers, "test", DATE);
        }
    }

    return parent;
}

function writeOutput(root) {
    const dir = path.dirname(OUTPUT_FILE);
    fs.mkdirSync(dir, { recursive: true });

    // Mirror Python: write explicit XML declaration (utf-8, newline) then the XML body
    const metatag = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n";
    const xmlBody = root.doc().end({ headless: true, prettyPrint: false });
    fs.writeFileSync(OUTPUT_FILE, metatag + xmlBody, "utf8");
}

function main() {
    const root = readFileAndBuildXml();
    writeOutput(root);
    console.log(`Wrote PAXML to ${OUTPUT_FILE}`);
}

// Execute if run directly
if (import.meta.url === `file://${process.argv[1]}`) {
    main();
}


